lib/mutgen.c

```diff
@@ -183,69 +198,94 @@ mutgen_init_allocator(mutgen_t *self, tsk_table_collection_t *tables)
 
 static int MSP_WARN_UNUSED
 mutgen_add_mutation(mutgen_t *self, node_id_t node, double position,
-        const char *ancestral_state, const char *derived_state)
+        const char *ancestral_state, const char *derived_state,
+        avl_node_t* avl_node)
 {
     int ret = 0;
-    tsk_site_t *site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+    bool new_site = (avl_node == NULL);
+    site_node_t *site_node;
+    tsk_site_t *site;
+    mutation_node_t *mutation_node = tsk_blkalloc_get(&self->allocator, sizeof(*mutation_node));
     tsk_mutation_t *mutation = tsk_blkalloc_get(&self->allocator, sizeof(*mutation));
-    avl_node_t* avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
-
-    if (site == NULL || mutation == NULL || avl_node == NULL) {
+    if (mutation == NULL || mutation_node == NULL) {
         ret = MSP_ERR_NO_MEMORY;
         goto out;
     }
-    memset(site, 0, sizeof(*site));
+    if (new_site) {
+        avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
+        site_node = tsk_blkalloc_get(&self->allocator, sizeof(*site_node));
+        site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+        if (site == NULL || site_node == NULL || avl_node == NULL) {
+            ret = MSP_ERR_NO_MEMORY;
+            goto out;
+        }
+        memset(site_node, 0, sizeof(*site_node));
+        site_node->site = site;
+        site_node->mutations_head = NULL;
+        site_node->mutations_length = 0;
+
+        memset(site, 0, sizeof(*site));
+        site->position = position;
+        site->ancestral_state = ancestral_state;
+        site->ancestral_state_length = 1;
+
+        avl_init_node(avl_node, site_node);
+        avl_node = avl_insert_node(&self->sites, avl_node);
+        assert(avl_node != NULL);
+    } else {
+        site_node = (site_node_t *) avl_node->item;
+        site = site_node->site;
+        assert(site->position == position);
+    }
+
     memset(mutation, 0, sizeof(*mutation));
-    site->position = position;
-    site->ancestral_state = ancestral_state;
-    site->ancestral_state_length = 1;
     mutation->derived_state = derived_state;
     mutation->derived_state_length = 1;
     mutation->node = node;
     mutation->parent = TSK_NULL;
-    site->mutations = mutation;
-    site->mutations_length = 1;
 
-    avl_init_node(avl_node, site);
-    avl_node = avl_insert_node(&self->sites, avl_node);
-    assert(avl_node != NULL);
+    memset(mutation_node, 0, sizeof(*mutation_node));
+    mutation_node->mutation = mutation;
+    mutation_node->next = (struct mutation_node_t *) site_node->mutations_head;
```

Comment:
Shouldn't need the cast here.

Reply:
The compiler gives me warnings otherwise, like:
```
lib/mutgen.c:246:25: warning: assignment to ‘struct mutation_list_t *’ from incompatible pointer type ‘mutation_list_t *’ {aka ‘struct <anonymous> *’} [-Wincompatible-pointer-types]
     mutation_list->next = site_list->mutations_head;
```